From 3f597ce2bd0babfb387cafce9805692565cbee8a Mon Sep 17 00:00:00 2001
From: Anakin Childerhose <anakin@childerhose.ca>
Date: Tue, 20 Jan 2026 21:47:08 -0500
Subject: [PATCH 1/2] add support for stopping the server

---
 init.lua | 23 +++++++++++++++++------
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git a/init.lua b/init.lua
index a4ffe2f..4f31a70 100644
--- a/init.lua
+++ b/init.lua
@@ -104,22 +104,34 @@ end
 ---@param callback fun(request: table, response: table)
 function Pegasus:start(callback)
   local handler = Handler:new(callback, self.location, self.plugins, self.log)
-  local server = assert(socket.bind(self.host, self.port))
-  local ip, port = server:getsockname()
+  self.server = assert(socket.bind(self.host, self.port))
+  local ip, port = self.server:getsockname()

   print('Pegasus is up on ' .. ip .. ":".. port) -- needed in case no LuaLogging is available
   handler.log:info('Pegasus is up on %s:%s', ip, port)

-  while 1 do
-    local client, errmsg = server:accept()
+  while self.server do
+    local client, errmsg = self.server:accept()

     if client then
       client:settimeout(self.timeout, 'b')
-      handler:processRequest(self.port, client, server)
+      handler:processRequest(self.port, client, self.server)
     else
-      handler.log:error('Failed to accept connection: %s', errmsg)
+      if errmsg ~= "timeout" then
+        handler.log:error('Failed to accept connection: %s', errmsg)
+      end
     end
   end
 end

+function Pegasus:stop()
+  print("stopping Pegasus")
+  if not self.server then return end
+  self.server:close()
+  self.server = nil
+  if self.log then
+    self.log:info('Pegasus server stopped')
+  end
+end
+
 return Pegasus
--
2.52.0

